src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java

```diff
@@ -386,80 +396,102 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
-                        c.getProjectName() != null
-                        && c.getProjectName().equals(expandedProjectName))
-                    .collect(Collectors.toList());
-            if(projectBuildActions == null || projectBuildActions.isEmpty()) {
-                csLogger.writeInfo(
-                    "Ignoring build since it has no matching build data. build=\"{0}\", project=\"{1}\"",
-                    previousBuildName,
-                    expandedProjectName);
-            } else if (projectBuildActions.size() > 1) {
-                csLogger.writeInfo(
-                    "Ignoring build since it has too many matching build actions. build=\"{0}\", project=\"{1}\", matches={2}",
-                    previousBuildName,
-                    expandedProjectName,
-                    projectBuildActions.size());
-            } else {
-                CodeSonarBuildAction previousBuildAction = projectBuildActions.get(0);
-                CodeSonarBuildActionDTO previousDTO = previousBuildAction.getBuildActionDTO();
-                URI previousBuildBaseHubUri = previousDTO.getBaseHubUri();
-                if(!baseHubUri.equals(previousBuildBaseHubUri)) {
-                    // TODO: we could try signing-in to the previous build hub;
-                    //  if it works, then perhaps we can still compare results?
+        // If specified, use parameter "base Analysis" as the analysis being used for comparison
+        if(StringUtils.isNotBlank(baseAnalysis)) {
+            Long baseAnalysisId = null;
+            try {
+                baseAnalysisId = Long.valueOf(baseAnalysis);
+            } catch(NumberFormatException e) {
+                LOGGER.log(Level.WARNING, "Unable to parse base analysis \"" + baseAnalysis + "\" as long integer.");
+            }
+            
+            previousDataLoader = new CodeSonarHubAnalysisDataLoader(
+                    httpService,
+                    hubInfo,
+                    baseHubUri,
+                    baseAnalysisId,
+                    getVisibilityFilterOrDefault(),
+                    getNewWarningsFilterOrDefault());
+            csLogger.writeInfo(
+                "Using analysis {0} on hub {1} for comparison",
+                String.valueOf(baseAnalysisId),
+                baseHubUri);
+        } else {
+            // Search for a previous, successful build that has the same ProjectName and hub URI.
+            //  We match hub URI since generally it is not reliable to compare results between different hubs.
+            //  Also, even if we want to consider other hubs, we may still be unable to communicate with them
+            //   due to CA certificate and user credential differences.
+            csLogger.writeInfo("Finding previous builds for comparison...");
+            Run<?,?> previousRun = run.getPreviousSuccessfulBuild();
+            while(previousRun != null 
+                && previousDataLoader == null
+            ) {
```

Comment:
Can we move this giant `else` clause containing the previous build search code to a dedicated method?  It seems complicated and procedural enough to deserve its own method.